File: platform/phonefarm/net.py
"""net — política central de egress/SSRF (paso 9).

- `guard_internal_url`: solo esquema http/https y host dentro de una allowlist
  explícita (MPT/Flask internos); rechaza redirects que cambien de host/esquema.
- `guard_external_url`: para destinos externos (LLM, Pexels): esquema http/https,
  sin credenciales en URL y las IPs resueltas NO pueden ser privadas/loopback/
  link-local/metadata (SSRF por DNS rebinding).

Toda salida HTTP de la plataforma debería pasar por aquí.
"""

from __future__ import annotations

import ipaddress
import logging
import socket
from urllib.parse import urlparse

import requests

logger = logging.getLogger(__name__)

# Hosts internos permitidos (MPT local/Docker; Flask loopback)
ALLOWED_INTERNAL_HOSTS = {"127.0.0.1", "localhost", "moneyprinter", "host.docker.internal"}

BLOCKED_IP_KINDS = ("is_private", "is_loopback", "is_link_local", "is_reserved", "is_multicast", "is_unspecified")


class EgressError(ValueError):
    """Destino no permitido por la política de egress."""


def _host_of(url: str) -> str:
    try:
        return (urlparse(url).hostname or "").lower()
    except ValueError as exc:
        raise EgressError("URL mal formada") from exc


def _resolve(host: str) -> list[str]:
    try:
        return sorted({info[4][0] for info in socket.getaddrinfo(host, None, socket.AF_UNSPEC)})
    except socket.gaierror as exc:
        raise EgressError(f"no se pudo resolver {host}") from exc


def _ip_blocked(ip_str: str) -> bool:
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        return True
    return any(getattr(ip, kind, False) for kind in BLOCKED_IP_KINDS)


def guard_internal_url(url: str, allowed_hosts: set[str] | None = None) -> None:
    """Permite solo http/https hacia hosts internos de la allowlist."""
    allowed = allowed_hosts or ALLOWED_INTERNAL_HOSTS
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        raise EgressError(f"esquema no permitido: {parsed.scheme!r}")
    if parsed.username or parsed.password:
        raise EgressError("credenciales en URL no permitidas")
    host = parsed.hostname or ""
    if host.lower() not in allowed:
        raise EgressError(f"host interno no permitido: {host}")


def guard_external_url(url: str) -> None:
    """Permite http/https a destinos externos SIN IPs privadas/metadata."""
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        raise EgressError(f"esquema no permitido: {parsed.scheme!r}")
    if parsed.username or parsed.password:
        raise EgressError("credenciales en URL no permitidas")
    host = parsed.hostname or ""
    if not host:
        raise EgressError("URL sin host")
    for ip in _resolve(host):
        if _ip_blocked(ip):
            raise EgressError(f"IP bloqueada ({ip}) para {host}")


def safe_get(url: str, *, internal: bool = True, allowed_hosts: set[str] | None = None,
             headers: dict[str, str] | None = None, timeout: float = 10,
             allow_redirects: bool = False, stream: bool = False, **kwargs) -> requests.Response:
    """GET con política de egress + redirects restringidos al mismo host/esquema."""
    guard_internal_url(url, allowed_hosts) if internal else guard_external_url(url)
    resp = requests.get(url, headers=headers, timeout=timeout, stream=stream,
                        allow_redirects=False, **kwargs)
    if resp.is_redirect and not allow_redirects:
        location = resp.headers.get("Location", "")
        if not location.startswith(("http://", "https://")):
            raise EgressError("redirect relativo no permitido")
        if internal:
            guard_internal_url(location, allowed_hosts)
        else:
            guard_external_url(location)
    return resp


def safe_post(url: str, *, internal: bool = True, allowed_hosts: set[str] | None = None,
              json=None, headers: dict[str, str] | None = None, timeout: float = 15, **kwargs) -> requests.Response:
    guard_internal_url(url, allowed_hosts) if internal else guard_external_url(url)
    return requests.post(url, json=json, headers=headers, timeout=timeout, allow_redirects=False, **kwargs)

File: server/net.ts
// ---------------------------------------------------------------------------
// Política central de egress/SSRF de Express (paso 9).
// - interno: solo http/https hacia hosts de la allowlist (loopback/MPT/Flask).
// - externo: sin credenciales en URL; resolución DNS que caiga en IP privada/
//   loopback/link-local se rechaza (anti SSRF/DNS rebinding).
// ---------------------------------------------------------------------------

import { lookup } from "dns/promises";
import net from "net";

export const INTERNAL_HOSTS = new Set(["127.0.0.1", "localhost", "moneyprinter", "host.docker.internal"]);

export class EgressError extends Error {}

function parse(url: string): URL {
  try {
    return new URL(url);
  } catch {
    throw new EgressError("URL mal formada");
  }
}

function isBlockedIp(ip: string): boolean {
  if (net.isIP(ip) === 0) return true;
  const parts = ip.split(".").map(Number);
  if (parts.length === 4) {
    if (parts[0] === 10) return true;
    if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31) return true;
    if (parts[0] === 192 && parts[1] === 168) return true;
    if (parts[0] === 127) return true;
    if (parts[0] === 169 && parts[1] === 254) return true; // link-local
    if (parts[0] === 0 || parts[0] >= 224) return true; // unspecified/multicast/reserved
  }
  if (ip.includes(":")) {
    const v = ip.toLowerCase();
    if (v === "::1" || v.startsWith("fe80") || v.startsWith("fc") || v.startsWith("fd") || v === "::") return true;
  }
  return false;
}

export function guardInternalUrl(url: string, allowedHosts: Set<string> = INTERNAL_HOSTS): URL {
  const u = parse(url);
  if (u.protocol !== "http:" && u.protocol !== "https:") throw new EgressError("esquema no permitido");
  if (u.username || u.password) throw new EgressError("credenciales en URL no permitidas");
  const host = u.hostname.toLowerCase();
  if (!allowedHosts.has(host)) throw new EgressError(`host interno no permitido: ${host}`);
  return u;
}

export async function guardExternalUrl(url: string): Promise<URL> {
  const u = parse(url);
  if (u.protocol !== "http:" && u.protocol !== "https:") throw new EgressError("esquema no permitido");
  if (u.username || u.password) throw new EgressError("credenciales en URL no permitidas");
  const host = u.hostname.toLowerCase();
  if (net.isIP(host) !== 0) {
    if (isBlockedIp(host)) throw new EgressError(`IP bloqueada: ${host}`);
    return u;
  }
  const records = await lookup(host, { all: true });
  for (const r of records) {
    if (isBlockedIp(r.address)) throw new EgressError(`IP bloqueada (${r.address}) para ${host}`);
  }
  return u;
}

/** GET interno seguro: misma política + redirects que cambien host/esquema → rechazo. */
export async function safeFetchInternal(url: string, init: RequestInit = {}, allowedHosts: Set<string> = INTERNAL_HOSTS): Promise<Response> {
  guardInternalUrl(url, allowedHosts);
  const res = await fetch(url, { ...init, redirect: "manual" });
  if (res.status >= 300 && res.status < 400) {
    const location = res.headers.get("location") || "";
    if (location.startsWith("http")) guardInternalUrl(location, allowedHosts);
  }
  return res;
}
